docs/installation/database/passing-configuration.rst

```diff
@@ -0,0 +1,430 @@
+side by side:
+
+::
+
+   dc_main    = Datacube(env="main")
+   dc_aux     = Datacube(env="aux")
+   dc_private = Datacube(env="private")
+
+3b. Specifying in the CLI
++++++++++++++++++++++++++
+
+The active environment can be selected in Python with the ``-E`` or ``--env`` option to the ``datacube``
+CLI tool.
+
+CLI commands that require more than one environment will have a second option for the second argument.
+Refer to the ``--help`` text for more information.
+
+3c. Via an Environment Variable
++++++++++++++++++++++++++++++++
+
+.. envvar:: ODC_ENVIRONMENT
+
+   If not explicitly specified via methods 3a. and 3b. above, the active
+   environment can be specified with the ``$ODC_ENVIRONMENT`` environment
+   variable.
+
+3d. Default Environment
++++++++++++++++++++++++
+
+If not specified by any of the methods 3a. to 3d. above, the ``default``
+environment is used.  If no ``default`` environment is known, an error is
+raised.  It is strongly recommended that a ``default`` environment be defined
+in all configuration files - either explicitly, or as an alias to an explicitly
+defined environment.
+
+If no environment named ``default`` is known, but one named ``datacube`` **IS**
+known, the ``datacube`` environment is used and a deprecation warning issued.
+``datacube`` will be dropped as a legacy default environment name in a future
+release.
+
+4. Generic Environment Variable Overrides
+-----------------------------------------
+
+Configuration values in config files can be over-ridden by setting the appropriate environment variable.
+
+The name of overriding environment variables are all upper-case and structured:
+
+.. code-block:: bash
+
+   $ODC_{environment name}_{option name}
+
+E.g. to override the :confval:`db_password` field in the ``main`` environment,
+set the ``$ODC_MAIN_DB_PASSWORD`` environment variable.
+
+Environment variables overrides are **NOT** applied to environments defined in
+configuration that was passed in explicitly as a string or dictionary.
+
+4a. Dynamic Environments
+++++++++++++++++++++++++
+
+It is possible for environments to be defined dynamically purely in environment variables.
+
+E.g. given the following active configuration file:
+
+.. code-block::yaml
+
+     default:
+         alias: main
+     main:
+         index_driver: postgres
+         db_url: postgresql://myuser:mypassword@server.domain/main
+
+and the following defined environment variables:
+
+.. code-block::bash
+
+   ODC_AUX_INDEX_DRIVER=postgis
+   ODC_AUX_DB_URL=postgres://auxuser:secret@backup.domain/aux
+
+You can request the "aux" environment and it's configuration will be
+dynamically read from the environment variables, even though it is not
+mentioned in the configuration file at all.
+
+Notes:
+
+1. Environment variables are read when configuration is first read from that
+   environment (i.e. when first connecting to the database.)
+
+2. As all configuration options have default values, it is no longer possible
```

Comment:
I think the main point of  this bullet is that environments are dynamically created, so lead the bullet with the "An all-defaults environment is used when requesting an environment that does not exist."

Reply:
Ooh nice. Yes.  Even better, cut straight to the bullet.  That's all just waffle.